Search: Add inline tags

psealock opened the pull request:
The Search component sometimes requires Tags to be rendered within the input instead of below. This PR creates a prop `inlineTags` to render tags inline. 

## Screenshots

![screen shot 2018-10-11 at 5 28 24 pm](https://user-images.githubusercontent.com/1922453/46780848-20d06580-cd7b-11e8-9ce0-b2c7072c4dd1.png)


## Test

1. If you have devDocs enabled, head to `/wp-admin/admin.php?page=wc-admin#/devdocs/search`. Otherwise, `/wp-admin/admin.php?page=wc-admin#/analytics/orders?filter=advanced` and add a Product filter.
2. Use the Search component to add Products
3. Is the behaviour expected? Be sure to use the keyboard to navigate as well. 

## Other Changes

1. Added a `<label/>` for the Search input.
2. Ran `npm run docs`

psealock committed:
Search: Add inline tags

psealock commented on .eslintrc:
```diff
@@ -29,6 +29,9 @@
+        "wpcalypso/redux-no-bound-selectors": 1,
+		"jsx-a11y/label-has-for": [ "error", {
+			"required": "id"
+		} ]
```
Following Gutenberg's lead on this:
https://github.com/WordPress/gutenberg/blob/80ac62507567a7e8c2f02693bbb05edc1532407b/eslint/config.js#L44-L46

timmyc commented on client/components/search/example.md:
```diff
@@ -4,11 +4,26 @@ import { Search } from '@woocommerce/components';
+				onChange={ items => setState( { selected: items } ) }
+			/>
+		</Section>
+		<H>Tags Inline with Input</H>
```
Nice! Thanks for updating the docs.

timmyc commented on docs/components/filters.md:
```diff
@@ -85,6 +85,63 @@ The `path` parameter supplied by React-Router.
 The query string represented in object form.
 
+`CompareFilter` (component)
+===========================
```
Thanks for re-generating the docs.

timmyc commented on client/components/search/index.js:
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
+								{ this.renderTags() }
+								<input
+									ref={ this.input }
+									type="text"
```
why is this type=text here and type=search below?

psealock replied:
"search" tells the browser to render the clear button on the right:

![screen shot 2018-10-12 at 9 59 52 am](https://user-images.githubusercontent.com/1922453/46833804-22933b00-ce06-11e8-972b-bd5080e05b70.png)

I don't think it makes sense in this context to have a clear button. Would it clear the text? Or all tags?

timmyc replied:
Ah gotcha, no I think **not** having the browser clear button makes sense for inline tags as it would likely cause confusion.

timmyc reviewed (commented):
Great stuff Paul. Couple small questions in the code, but also wanted to see if @ryelle wanted to 👀 this one too due to her prior work on this component. From testing things out in Dev Docs, I noticed the page would jump around a bit as I worked with the components... typically after selecting the first product, or removing the last product:

![scroll-jump](https://user-images.githubusercontent.com/22080/46801308-2abc7f00-cd0f-11e8-8407-810c0086d14f.gif)

Not sure if this is due to having the two instances on the same page using the same data, but we can just watch out for this as we integrate it in reports and handle it in a future PR if it is persistent.

The last thing I noticed with the inline tags is later in the .gif above, the div appears to get quite a bit of padding on the bottom when longer product names are added. Not a massive deal but figured I'd point it out too.

ryelle commented on client/components/search/index.js:
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
-									screenReaderLabel={ screenReaderLabel }
+					{ ( { listBoxId, activeId, onChange } ) =>
+						// Disable reason for no-static-element-interactions: There is no appropriate role
+						// which describes the input container intended accessible usability.
```
Can you rephrase this a little? Actually you could probably merge both disable reasons into one, since it's really the same for both, right?

ryelle commented on client/components/search/index.js:
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
+					const screenReaderLabel = sprintf(
+						__( '%1$s (%2$s of %3$s)', 'wc-admin' ),
+						item.label,
+						i,
```
`i` should be `i+1`, the index is zero-based so we get "T-Shirt (0 of 1)"

ryelle commented on client/components/search/index.js:
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
 		return (
 			<div className="woocommerce-search">
+				<label htmlFor={ `search-input-${ instanceId }` } className="screen-reader-text">
+					{ placeholder || __( 'Add item', 'wc-admin' ) }
```
We're already using `aria-label` or `aria-labelledby` to fill this purpose, we don't actually need to add a `label` here (especially an invisible label).

psealock replied:
Oh right of course. Those are passed in via props. How do you feel about making them `required`?

ryelle replied:
Well, you wouldn't have both– `aria-label` would just be a string, and `aria-labelledby` is an element ID. If you want to have a required prop, we could change it to accept a `label` prop, which would be a string, and we can either pass it to the input's `aria-label` or this `<label>` element. We'd have to change how the `SearchFilter` works, though (that uses `ariaLabelledby`, which is actually not right).

psealock replied:
>We'd have to change how the SearchFilter works, though (that uses ariaLabelledby, which is actually not right).

This reminds me, I've changed this in https://github.com/woocommerce/wc-admin/pull/457 to use `aria-label` as suggested

ryelle commented on client/components/search/index.js:
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
+									aria-activedescendant={ activeId }
+									onFocus={ this.onFocus }
+									onBlur={ this.onBlur }
+									{ ...aria }
```
It might be good to add an `aria-describedby` div that said "Move [forward/backward] for selected items" or something like that, so screen reader users know where these show up

ryelle reviewed (commented):
A few notes, but looks good 👍 I don't see the jumping issue @timmyc  mentioned, though.

psealock commented:
Thanks for the review @timmyc @ryelle 

* Changed `state` in the example.md so the jumpiness isn't there when looking at devDocs
* Removed the unneeded `<label>` 
* Updated the eslint disabled comment to be more readable
* Fixed item label numbering issue
* Added `aria-describedby` when items have been selected so screen readers can tell the user where selected items can be found

timmyc reviewed (approved):
Thanks for the updates and the responses on my questions. Testing well for me and looking splendid too
